rigging/generator/http.py

```diff
@@ -0,0 +1,285 @@
+from __future__ import annotations
+
+import asyncio
+import base64
+import json
+import re
+import typing as t
+
+import httpx
+import jinja2
+import jsonpath_ng  # type: ignore
+from pydantic import BaseModel, ConfigDict, Field, field_serializer, field_validator
+from ruamel.yaml import YAML
+
+from rigging.error import ProcessingError, raise_as
+from rigging.generator.base import (
+    GeneratedMessage,
+    GenerateParams,
+    Generator,
+    trace_messages,
+)
+from rigging.message import Message, Role
+
+# TODO:
+# - Add request retry mechanics
+# - Add request timeout mechanics
+# - Add maximum concurrent requests
+
+# Helpers
+
+
+def _to_str(v: str | dict[str, t.Any]) -> str:
+    return json.dumps(v) if isinstance(v, dict) else v
+
+
+def _to_dict(v: str | dict[str, t.Any]) -> dict[str, t.Any]:
+    return t.cast(dict[str, t.Any], json.loads(v)) if isinstance(v, str) else v
+
+
+def _to_dict_or_str(v: str) -> dict[str, t.Any] | str:
+    try:
+        return t.cast(dict[str, t.Any], json.loads(v))
+    except json.JSONDecodeError:
+        return v
+
+
+# Jinja/template context when building request bodies
+
+
+class RequestTransformContext(BaseModel):
+    role: Role
+    """Role of the last message in the sequence."""
+
+    content: str
+    """Content of the last message in the sequence."""
+
+    all_content: str
+    """Concatenation of all message content in the sequence."""
+
+    messages: list[dict[str, t.Any]]
+    """List of all messages objects in the sequence (.role, .content)."""
+
+    params: dict[str, t.Any]
+    """Merged parameters of the incoming request and generator params."""
+
+    api_key: str
+    """API key set on the generator."""
+
+    model: str
+    """Model set on the generator."""
+
+
+# Spec types
+
+InputTransform = t.Literal["json", "jinja"]
```

Comment:
should the input also include `jsonpath`?

Reply:
I think of `jsonpath` as being an extractor mechanism (removing information), and `json` as formatting (adding information).

Def open to some example use cases though, lots of options for blending.